### ISSUE TITLE
SNOW-782588: Long running async queries are not handled correctly

### ISSUE THREAD

**Issue by pravindra** [snowflakedb/gosnowflake#768]
Please answer these questions before submitting your issue. 
In order to accurately debug the issue this information is required. Thanks!

1. What version of GO driver are you using?

1.6.16
  
2. What operating system and processor architecture are you using?

Mac, arm64
   
3. What version of GO are you using?
run `go version` in your console

go1.19.4 darwin/arm64

4.Server version:* E.g. 1.90.1
You may get the server version by running a query:
```
SELECT CURRENT_VERSION();
```

7.10.1

5. What did you do?

   If possible, provide a recipe for reproducing the error.
   A complete runnable program is good.

Ran an async query that takes longer than a minute.

```
ctx, _ := sf.WithMultiStatement(context.Background(), 0)
rows, err := db.QueryContext(sf.WithAsyncMode(ctx), "CALL SYSTEM$WAIT(1, 'MINUTES');use snowflake_sample_data")
if err != nil {
	log.Fatalf("failed to run a query. %v, err: %v", query, err)
}
defer rows.Close()
var v interface{}
for {
	for rows.Next() {
		err := rows.Scan(&v)
		if err != nil {
			log.Fatalf("failed to get result. err: %v", err)
		}
		log.Printf("result: %v", v)
	}
	if !rows.NextResultSet() {
		break
	}
}
```
6. What did you expect to see?

```
result: waited 1 minutes
result: Statement executed successfully.
```

   What should have happened and what happened instead?

did not get any output.

I got the correct behaviour with this patch

```
diff --git a/async.go b/async.go
index 5c477dc..80b704f 100644
--- a/async.go
+++ b/async.go
@@ -63,6 +63,7 @@ func (sr *snowflakeRestful) getAsync(
        defer close(errChannel)
        token, _, _ := sr.TokenAccessor.GetTokens()
        headers[headerAuthorizationKey] = fmt.Sprintf(headerSnowflakeToken, token)
+RETRY:
        resp, err := sr.FuncGet(ctx, sr, URL, headers, timeout)
        if err != nil {
                logger.WithContext(ctx).Errorf("failed to get response. err: %v", err)
@@ -86,6 +87,9 @@ func (sr *snowflakeRestful) getAsync(
 
        sc := &snowflakeConn{rest: sr, cfg: cfg}
        if respd.Success {
+               if respd.Code == queryInProgressCode || respd.Code == queryInProgressAsyncCode {
+                       goto RETRY
+               }
                if resType == execResultType {
                        res.insertID = -1
                        if isDml(respd.Data.StatementTypeID) {

```
7. Can you set logging to DEBUG and collect the logs?

   https://community.snowflake.com/s/article/How-to-generate-log-file-on-Snowflake-connectors
   
8. What is your Snowflake account identifier, if any? (Optional)

**Comment by sfc-gh-dszmolka** [comment]
hi and thank you for submitting this issue with us ! `SYSTEM$WAIT` is special on the backend too so I would like to ask if you see this unexpected behaviour also when you run a 'normal' query, e.g. any actual SELECT, INSERT, etc which takes more than 1 minute ?

i'll try and reproduce the same but if you perhaps already tried that, that would be great if you could share your observations.

**Comment by pravindra** [comment]
sorry, i haven't tried with an actual select/insert.

**Comment by sfc-gh-dszmolka** [comment]
is it possible to try it please and see how it behaves ?

**Comment by pravindra** [comment]
I simulated the wait with a javascript UDF & was able to reproduce the same issue.

The javascript UDF (it does a sleep of 1s for each input)

```
CREATE OR REPLACE FUNCTION SLEEP_FN(STR VARCHAR)
  RETURNS VARCHAR
  LANGUAGE JAVASCRIPT
  AS $$
    var date = new Date();
    var curDate = null;
    do { curDate = new Date(); }
    while(curDate-date < 1000);
  
    return STR + ":" + date.toString();
  $$
  ;
```

I then invoked this fn with two select queries that have 25 records each (so, query takes 50s to run)

```
ctx, _ := sf.WithMultiStatement(context.Background(), 0)
rows, err := db.QueryContext(sf.WithAsyncMode(ctx), "select sleep_fn(n_name) from snowflake_sample_data.tpch_sf1.nation;select sleep_fn(n_name) from snowflake_sample_data.tpch_sf1.nation;use snowflake_sample_data")
```

With the vanilla gosnowflake client, I get 0 records in the result. --> incorrect result
With my patch above, I get 51 records in the result. --> expected result

**Comment by sfc-gh-dszmolka** [comment]
thank you for checking! i'll check if the issue can be reproduced following the [async queries documentation](https://pkg.go.dev/github.com/snowflakedb/gosnowflake#hdr-Asynchronous_Queries) and involve the driver team if necessary. thank you for bearing with us.

**Comment by sfc-gh-dszmolka** [comment]
Thanks for the patience here ! Per the driver team, the execution/implementation for the async queries look like this.

The status will only change to QueryStatusComplete in waitForAsyncExecStatus() or waitForAsyncQueryStatus(), which are being called by the API functions (blocking calls) implemented in snowflakeRows and snowflakeResult.

The following functions in snowflakeResult call [waitForAsyncExecStatus](https://github.com/snowflakedb/gosnowflake/blob/v1.6.20/result.go#L61):

- LastInsertId
- RowsAffected

The following functions in snowflakeRows call [waitForAsyncQueryStatus](https://github.com/snowflakedb/gosnowflake/blob/v1.6.20/rows.go#L221):

- Close
- ColumnTypeDatabaseTypeName
- ColumnTypeLength
- ColumnTypeNullable
- ColumnTypePrecisionScale
- Columns
- ColumnTypeScanType
- GetArrowBatches
- Next
- HasNextResultSet
- NextResultSet

 By calling one of the above blocking calls, the result set will be filled when it is ready.
 
 Hope this helps. Also marking this one as closed for now as it seems to work as expected, but if you still feel you found a bug, please let us know and I'll reopen this issue.

**Comment by pravindra** [comment]
> Also marking this one as closed for now as it seems to work as expected, but if you still feel you found a bug, please let us know

if you see the repro case I pasted in my description, it makes calls to both Next and NextResultSet (these are two functions that you listed above). And, that still caused the bug.

Can you please look at the repro case in the description & suggest what should be changed to avoid the bug ?

**Comment by madisonchamberlain** [comment]
I have noticed the same behavior as above but with Close(). When Close is called on the results it still won't wait until the query is completed. We had to fork the driver and implement the same retry behavior as above to get Close() to block for the query to complete if submitted for async execution. From what I have seen, anything that calls `getQueryResultResp` under the hood will not work for longer than 45 seconds no matter what method is called on the results.
